Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.cpp

```diff
@@ -100,6 +100,78 @@ JSWebAssemblyArray::~JSWebAssemblyArray()
     }
 }
 
+void JSWebAssemblyArray::fill(uint32_t offset, EncodedJSValue value, uint32_t size)
+{
+    // Handle ref types separately to ensure write barriers are in effect.
+    if (isRefType(m_elementType.type.unpacked()) && !isI31ref(m_elementType.type.unpacked())) {
+        for (size_t i = 0; i < size; i++)
+            set(offset + i, value);
+        return;
+    }
+
+    if (m_elementType.type.is<Wasm::PackedType>()) {
+        switch (m_elementType.type.as<Wasm::PackedType>()) {
+        case Wasm::PackedType::I8:
+            memset(m_payload8.data() + offset, static_cast<uint8_t>(value), size);
+            return;
+        case Wasm::PackedType::I16:
+            std::fill(m_payload16.begin() + offset, m_payload16.begin() + offset + size, static_cast<uint16_t>(value));
+            return;
+        }
+    }
+
+    switch (m_elementType.type.as<Wasm::Type>().kind) {
+    case Wasm::TypeKind::I32:
+    case Wasm::TypeKind::F32:
+        std::fill(m_payload32.begin() + offset, m_payload32.begin() + offset + size, static_cast<uint32_t>(value));
+        return;
+    // FIXME: support v128
+    default:
+        std::fill(m_payload64.begin() + offset, m_payload64.begin() + offset + size, value);
```

Comment:
This is missing a write barrier, right? You should be able to get it to crash with collectContinuously=1

Reply:
This should be ok actually, because reftypes are handled separately. Specifically on line 106 if the element type is a reftype then it will actually call the `set` method on each index instead of doing `std::fill`, which will use the write barrier path.

Same with the `copy` method below which also has a slow path for reftypes. 

There might be a write barrier issue with `array.init_elem` though, which I'll look into before merging.

Reply:
For `array.init_elem`, I added a `vm.writeBarrier()` call to ensure the write barrier is used after the write into the array (though I wasn't able to trigger any failures with test cases for this). That should address everything that's left on this.